File: discipline_client_ts/discipline_android_primitive_ts/rules/countdown_rules.ts
import { CountdownRule, Nominal } from "../x.ts";

const BRAND = Symbol();

export type CountdownRules = Nominal<typeof BRAND, CountdownRule[]>;

export const construct = (rules: CountdownRule[]): CountdownRules => {
  return Nominal.create(BRAND, rules);
};

export const createDefault = (): CountdownRules => {
  return construct([]);
};

export const CountdownRules = {
  createDefault,
};

File: discipline4/app/src/main/java/com/example/app/rules/CountdownRules.kt
package com.example.app

import com.example.app.*

public data class CountdownRules(
  val rules: MutableMap<UuidV4, CountdownRule>,
) {
  companion object {
    fun createDefault(): CountdownRules {
      return CountdownRules(rules = mutableMapOf())
    }
  }

  fun isActive(now: Instant): Boolean {
    return rules.values.any { rule -> rule.isActive(now) }
  }

  fun has(id: UuidV4): Boolean {
    return rules.containsKey(id)
  }

  fun get(id: UuidV4): CountdownRule? {
    return rules.get(id)
  }

  fun add(id: UuidV4, rule: CountdownRule) {
    rules.set(id, rule)
  }

  fun remove(id: UuidV4) {
    rules.remove(id)
  }
}
